### PR TITLE
Added policy compliant component selection (PCCS) for NPM CLM-21924

### PR DESCRIPTION
**Pull request by vladt**
https://issues.sonatype.org/browse/CLM-21924
https://jenkins.ci.sonatype.dev/job/insight/job/firewall-for-artifactory/job/feature-snapshots/job/CLM-21924_PCCSForNPM/

### CODE REVIEW

File: firewall/src/main/java/com/sonatype/iq/artifactory/NpmComponentMetadataProcessor.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        internalRepositoryServiceClass.getMethod("remoteRepositoryByKey", String.class);
+    return remoteRepositoryByKeyMethod.invoke(internalRepositoryService, repoPath.getRepoKey());
+  }
+
+  private void closeStreamHandle(
+      Object streamHandle) throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException, IllegalAccessException
+  {
+    Method closeMethod = streamHandle.getClass().getDeclaredMethod("close");
+    closeMethod.invoke(streamHandle);
+  }
```
Comment (fernau) [inline firewall/src/main/java/com/sonatype/iq/artifactory/NpmComponentMetadataProcessor.java:184]: this is fine, except it's a bit hard to read with the 'logic' and the 'reflection' both being here together.

would this be easier to read if we create a proxy class for the InternalRepositoryService which deals with the reflection only, and provides proper methods you can call? Then all this reflection noise could be removed from the NpmComponentMetadataProcessor

Comment (vladt) [reply firewall/src/main/java/com/sonatype/iq/artifactory/NpmComponentMetadataProcessor.java:184]: > would this be easier to read if we create a proxy class for the InternalRepositoryService which deals with the reflection only, and provides proper methods you can call? Then all this reflection noise could be removed from the NpmComponentMetadataProcessor

I separated the reflection code into methods. I'm not sure what moving it to a different class would buy us. It wouldn't really be easier to understand. It would only hide it.

In other words:
If one is not interested in the implementation details, then they don't have to look into those methods. O don't look into that other class.
If one is interested in the implementation details, then they have look into those methods. Or look into that other class.

**Review by fernau — APPROVED** [review]
LGTM

**Review by DanielRuaMadrid — APPROVED** [review]
